File: packages/sense/src/swift.ts
/**
 * What a Swift file asks for and what it publishes.
 *
 * Swift is the language that does not have the edge this package is built on.
 * `import Core` names a **module**, and a module is a whole target — tens or
 * hundreds of files compiled together. Nothing in the statement, and nothing
 * anywhere in the file, says which of those files the symbol came from. Within a
 * target it is worse: files see each other with no import at all, so a Swift file
 * names none of the files it depends on and cannot be made to
 * ([ADR-0066](../../../docs/context/adr/0066-a-language-is-a-reader-not-a-sense.md)).
 *
 * ## So the target is the answer, projected onto files
 *
 * An `import` becomes an edge to **every file in that target**, and a file's own
 * target becomes edges to every file beside it. Both are true statements at the
 * grain the language actually has: a change to any file in `Core` can change
 * what `Core` publishes, so anything that imported `Core` may be affected by it.
 * The projection over-reaches in the way selection is allowed to over-reach —
 * a test that did not need running is cost, and a test that needed running and
 * was not run is the failure this package exists to refuse.
 *
 * Measured on a 708-file, 51-target SwiftPM package: the largest target is 78
 * files, and same-target visibility is 20,756 edges — about twenty-nine per
 * file. The shape is affordable because targets are small on purpose; a
 * repository that put everything in one target would pay for it, and so would
 * its compiler.
 *
 * ## `Package.swift` is a program, and is read as one
 *
 * There is no manifest format to parse. Targets are declared by calling
 * `.target(name:path:)` inside Swift source, the `path:` is frequently not the
 * default, and the default itself — `Sources/<name>` — is only a default. Since
 * a Swift grammar is already loaded to read every other file, the manifest is
 * read with it, and the conventional layout is the fallback for a manifest that
 * could not be read rather than the rule.
 *
 * ## What is unknown
 *
 * A parse error **where an import could have been**, and a missing grammar. A
 * grammar recovers from syntax it cannot read by wrapping it in an `ERROR` node
 * and carrying on, so one unreadable expression inside one function body costs
 * nothing: every import in the file is still there to be read. Marking the file
 * unknown over that would report a single unsupported form — one the grammar is
 * simply a version behind on — as imports nobody could read, when every one of
 * them was. An error hides an import
 * only when it sits where this reader looks, which for Swift is the top level,
 * because that is the only place an `import` is allowed.
 *
 * `@_exported import` re-exports a module through this one and is read as an
 * ordinary import, which under-reports the transitive reach by exactly one hop.
 */

import { native, nativeRefusal } from './native.js';
import type { TreeWorld } from './world.js';

/** The request a file makes for its own target: the files it sees without asking. */
const OWN_TARGET = '*';

const SWIFT_FILE = '.swift';
const MANIFEST = 'Package.swift';

/** Swift requests never name a path, so a Swift request is never a hole. */
export function isSwiftRelative(): boolean {
  return false;
}

/** Every file a Swift module name reaches, which is every file in its target. */
export function resolveSwift(input: {
  readonly from: string;
  readonly request: string;
  readonly world: TreeWorld;
}): readonly string[] {
  const { from, request, world } = input;
  // A manifest is a build description, not a member of what it describes. Left
  // in, it sits under no target, falls back to the repository root, and joins
  // every Swift file in the tree to every other one.
  if (isManifest(from)) return [];

  const targets = targetIndex(world);
  if (request !== OWN_TARGET) {
    const declared = targets.byName.get(request);
    return declared === undefined ? [] : sourcesUnder(world, declared, from, true);
  }

  const own = targets.directories.find((at) => from.startsWith(`${at}/`));
  if (own !== undefined) return sourcesUnder(world, own, from, true);

  // A Swift file no package claims — an Xcode project, a loose script, a target
  // declared somewhere this could not read. Its own directory is the narrowest
  // thing certainly true of it, and only the files beside it: without a
  // manifest saying so, nothing says a subdirectory compiles with it.
  const cut = from.lastIndexOf('/');
  return sourcesUnder(world, cut === -1 ? '' : from.slice(0, cut), from, false);
}

function isManifest(path: string): boolean {
  return path === MANIFEST || path.endsWith(`/${MANIFEST}`);
}

function sourcesUnder(
  world: TreeWorld,
  directory: string,
  from: string,
  deep: boolean,
): readonly string[] {
  const paths = deep ? world.below(directory) : world.under(directory);
  return paths.filter((path) =>
    path.endsWith(SWIFT_FILE) && path !== from && !isManifest(path));
}

interface Targets {
  /** Module name → the directory its sources sit under. */
  readonly byName: ReadonlyMap<string, string>;
  /** Every target directory, longest first, so the innermost wins. */
  readonly directories: readonly string[];
}

function targetIndex(world: TreeWorld): Targets {
  return world.index('swift:targets', (tree) => {
    const byName = new Map<string, string>();

    for (const manifest of tree.below('')) {
      if (manifest !== MANIFEST && !manifest.endsWith(`/${MANIFEST}`)) continue;
      const at = manifest.slice(0, Math.max(0, manifest.length - MANIFEST.length - 1));

      // The convention first, so a manifest that could not be read still
      // produces targets, and a declared `path:` overwrites the guess after.
      for (const kind of ['Sources', 'Tests']) {
        for (const path of tree.below(under(at, kind))) {
          const rest = path.slice(under(at, kind).length + 1);
          const slash = rest.indexOf('/');
          if (slash > 0) byName.set(rest.slice(0, slash), under(at, `${kind}/${rest.slice(0, slash)}`));
        }
      }

      for (const target of declaredTargets(tree.text(manifest))) {
        byName.set(target.name, under(at, target.path));
      }
    }

    return {
      byName,
      directories: [...new Set(byName.values())].sort((a, b) => b.length - a.length),
    };
  });
}

function under(at: string, rest: string): string {
  return at === '' ? rest : `${at}/${rest}`;
}

type Target = { readonly name: string; readonly path: string };

/**
 * The `.target(name:path:)` calls in a manifest, read with the addon's Swift grammar.
 *
 * Nothing when the addon was built without its grammars: the conventional
 * layout [`targetIndex`](#targetIndex) laid down first is then the whole answer,
 * as it is for a manifest that does not parse.
 */
function declaredTargets(text: string | undefined): readonly Target[] {
  if (text === undefined) return [];
  const addon = native();
  if (addon === undefined) {
    throw new Error(`sense: reading Package.swift needs the native addon, which did not load: ${nativeRefusal()}`);
  }
  const answer = addon.swiftTargets(text);
  return answer === null ? [] : JSON.parse(answer) as Target[];
}

File: packages/sense/native/src/languages/swift.rs
//! What a Swift file asks for and what it publishes.
//!
//! The argument for the shape is [`swift.ts`](../../../src/swift.ts), which this
//! is a port of and is measured against: `import Core` names a **target**, so an
//! import becomes an edge to every file in it and a file's own target becomes
//! edges to every file beside it. Resolution — the target index and the
//! projection onto paths — stays in TypeScript, because it reads a tree and not
//! a file. What is here is the reader, and [`targets`], which reads the one file
//! resolution needs parsed: `Package.swift`.

use std::collections::HashSet;

use serde::Serialize;
use tree_sitter::{Node, Tree};

use super::shape::{Binding, Export, Kind, Read, Request};
use super::{line_of, named_children, text_of};

/// The request a file makes for its own target: the files it sees without asking.
const OWN_TARGET: &str = "*";

pub fn read(file: &str, source: &str, tree: &Tree) -> Read {
    let root = tree.root_node();
    let mut requests: Vec<Request> = Vec::new();
    let mut seen: HashSet<String> = HashSet::new();
    let mut exports: Vec<Export> = Vec::new();
    let mut broken = false;

    // The files beside this one, which it sees with no statement of any kind.
    want(&mut requests, &mut seen, OWN_TARGET, OWN_TARGET, 1);

    for child in named_children(root) {
        let line = line_of(child);
        if child.kind() == "ERROR" || child.is_missing() {
            broken = true;
        }
        if child.kind() == "import_declaration" {
            if child.has_error() {
                broken = true;
            }
            // `import struct Answer.Lens` names the module `Answer`; the rest of
            // the path is a symbol inside it, and there is no file grain below it.
            if let Some(module) = imported(child, source) {
                want(&mut requests, &mut seen, &module, &module, line);
            }
            continue;
        }
        if let Some(name) = declared(child, source) {
            exports.push(Export {
                exported: Some(name.clone()),
                local: Some(name),
                type_only: false,
                line,
                ..Export::default()
            });
        }
    }

    Read {
        requests,
        exports: if exports.is_empty() { None } else { Some(exports) },
        unknown: broken.then(|| {
            format!("{file} did not parse cleanly as Swift, so what it imports may be incomplete.")
        }),
    }
}

fn imported(node: Node, source: &str) -> Option<String> {
    let name = named_children(node)
        .into_iter()
        .find(|part| part.kind() == "identifier")?;
    match named_children(name).first() {
        Some(first) => Some(text_of(*first, source).to_string()),
        None => Some(
            text_of(name, source)
                .split('.')
                .next()
                .unwrap_or("")
                .to_string(),
        ),
    }
}

fn want(requests: &mut Vec<Request>, seen: &mut HashSet<String>, value: &str, local: &str, line: u32) {
    if value.is_empty() || !seen.insert(value.to_string()) {
        return;
    }
    requests.push(Request {
        value: value.to_string(),
        kind: Kind::Imports,
        bindings: vec![Binding {
            imported: local.to_string(),
            local: local.to_string(),
            type_only: false,
            line,
        }],
        line,
        // Always. A module name is not a path, `Foundation` and `Core` are
        // written identically, and the target a name belongs to is a fact about
        // the tree.
        guessed: true,
    });
}

/// Top-level declarations this file publishes.
///
/// `class_declaration` is every nominal type in this grammar — `struct`, `class`,
/// `enum` and `actor` all reach it — and the keyword that separates them is an
/// anonymous node. Nothing here needs to tell them apart.
const DECLARATIONS: [&str; 5] = [
    "class_declaration",
    "protocol_declaration",
    "typealias_declaration",
    "function_declaration",
    "property_declaration",
];

fn declared(node: Node, source: &str) -> Option<String> {
    if !DECLARATIONS.contains(&node.kind()) {
        return None;
    }
    let children = named_children(node);
    let named = |nodes: &[Node]| -> Option<String> {
        nodes
            .iter()
            .find(|child| child.kind() == "type_identifier" || child.kind() == "simple_identifier")
            .map(|child| text_of(*child, source).to_string())
    };
    if let Some(name) = named(&children) {
        return Some(name);
    }
    // `let value = 1` binds through a pattern rather than naming itself.
    let pattern = children.iter().find(|child| child.kind() == "pattern")?;
    named_children(*pattern)
        .into_iter()
        .find(|child| child.kind() == "simple_identifier")
        .map(|child| text_of(child, source).to_string())
}

/// A target a manifest declares: its name, and the directory its files sit in.
#[derive(Serialize)]
pub struct Target {
    pub name: String,
    pub path: String,
}

/// The `.target(name:path:)` calls in a `Package.swift`, in source order.
///
/// Here rather than beside the resolver because the parse is what costs: Node 24
/// and 25 abort with a V8 zone overflow while optimizing the WebAssembly Swift
/// grammar, and a manifest was the last Swift file that grammar still read on a
/// machine with the addon.
pub fn targets(source: &str, tree: &Tree) -> Vec<Target> {
    let mut found = Vec::new();
    // Preorder, as the JavaScript walk visits, and with a stack rather than
    // recursion so a deeply nested manifest cannot take the thread with it.
    let mut stack = vec![tree.root_node()];
    while let Some(node) = stack.pop() {
        if node.kind() == "call_expression" {
            if let Some(target) = declared_target(node, source) {
                found.push(target);
            }
        }
        stack.extend(named_children(node).into_iter().rev());
    }
    found
}

fn declared_target(call: Node, source: &str) -> Option<Target> {
    let callee = callee_of(call, source)?;
    if !callee.to_lowercase().ends_with("target") {
        return None;
    }
    let args = named_children(call)
        .into_iter()
        .find(|child| child.kind() == "call_suffix")
        .and_then(|suffix| {
            named_children(suffix)
                .into_iter()
                .find(|child| child.kind() == "value_arguments")
        });
    let name = argument(args, "name", source)?;
    // SwiftPM's own defaults, which are what a target with no `path:` means.
    let path = argument(args, "path", source).unwrap_or_else(|| {
        if callee == "testTarget" {
            format!("Tests/{name}")
        } else {
            format!("Sources/{name}")
        }
    });
    Some(Target { name, path })
}

/// `.target` parses as a prefix expression: the dot is the prefix and the name follows.
fn callee_of<'source>(call: Node, source: &'source str) -> Option<&'source str> {
    let head = named_children(call).into_iter().next()?;
    match head.kind() {
        "simple_identifier" => Some(text_of(head, source)),
        "prefix_expression" => named_children(head)
            .into_iter()
            .find(|child| child.kind() == "simple_identifier")
            .map(|child| text_of(child, source)),
        _ => None,
    }
}

fn argument(args: Option<Node>, label: &str, source: &str) -> Option<String> {
    for value in named_children(args?) {
        if value.kind() != "value_argument" {
            continue;
        }
        let children = named_children(value);
        let written = children
            .iter()
            .find(|child| child.kind() == "value_argument_label")
            .map(|child| text_of(*child, source));
        if written != Some(label) {
            continue;
        }
        let literal = children
            .iter()
            .find(|child| child.kind() == "line_string_literal")?;
        return Some(
            named_children(*literal)
                .into_iter()
                .map(|part| text_of(part, source))
                .collect(),
        );
    }
    None
}
